internal/cephfs/core/volume.go

```diff
@@ -240,7 +247,7 @@ func (s *subVolumeClient) CreateVolume(ctx context.Context) error {
 	}
 
 	// create subvolumegroup if not already created for the cluster.
-	if !clusterAdditionalInfo[s.clusterID].subVolumeGroupsCreated[s.FsName] {
+	if !s.isSubVolumeGroupCreated(s.SubvolumeGroup) {
```

Comment:
This is still racy. two or more go routines can call the function, and both can continue to create the SubVolumeGroup.

In order to prevent races, you will need to take a lock on the `clusterAdditionalInfo[s.clusterID]` object (a single lock for all members to keep it simple would be fine), or

1. lock the `.subVolumeGroupsRWMutex`
2. create the SubVolumeGroup
3. unlock the `.subVolumeGroupsRWMutex`

If access to the different members needs similar serialization, use a single Mutex to update any of the members and have other go routines block until the updating is done.

Reply:
> This is still racy. two or more go routines can call the function, and both can continue to create the SubVolumeGroup.
> 
> In order to prevent races, you will need to take a lock on the `clusterAdditionalInfo[s.clusterID]` object (a single lock for all members to keep it simple would be fine), or
> 
> 1. lock the `.subVolumeGroupsRWMutex`
> 2. create the SubVolumeGroup
> 3. unlock the `.subVolumeGroupsRWMutex`
> 
> If access to the different members needs similar serialization, use a single Mutex to update any of the members and have other go routines block until the updating is done.

Yes that is intentional, 
The mentioned parallel create will happen only once and there's no side affects to create being called more than once. The follow on update will still be guarded by rwlock so we're fine. 
I don't want to hold locks while we wait for a response from ceph.
And I want don't want other members to wait while one is being updated.

Reply:
Sorry, but that is not what the commit message describes:

```
Multiple go-routines may simultaneously check for
presence of a clusterID's metadata such as
subvolumegroup created state, resize state,
metadata state and snapshot metadata state in the
clusterAdditionalInfo map and update an entry after creation
if it is absent. This set of operation needs to be serialized.
```

There is no serialization in this commit. This commit add atomically setting/reading a value. Usually an int is atomic already, and sufficient if there are no calculations done with it. I do not think the mutex per member in the `operationState` contributes to preventing race conditions.